### ISSUE TITLE
How to prevent Steamworks exploitation?

### ISSUE THREAD

**Issue by DarklightStudios** [Facepunch/Facepunch.Steamworks#559]
Hi,

So I've been working with the Steamworks SDK for a few months now and finally have my own App ID (good bye Spacewar). I've been wondering how we would do things from the Unity client like update stats, add/remove/use items etc... But, securely.

I've noticed that if you're in the Unity Editor, you can unlock all the achievements for a game you own just by using the App ID through Unity. I'm not sure if this can be prevented for achievements as I accidently did this in a pretty big Steam game... Won't mention the name.

I see in the Steam backend you can set your leaderboards to only accept read/writes from the WebAPI, but don't know how that would work with Unity?

Long story short, how would I prevent someone from opening a Unity Editor, adding my game's App ID and giving themselves what ever stats/items they want, while also being able to legitimately update these things from the client?

Thanks in advance,
Michael.

**Comment by bobbahbrown** [comment]
This question may be better asked in the Steamworks Developers group on STEAM, you might get some better ideas there.

**Comment by SupinePandora43** [comment]
you shouldn't try to prevent Achievements & Stats abuse.
because it's impossible.
the **SAM** can easily do those things for free and open source.

**Comment by DarklightStudios** [comment]
Thanks, I've decided to rather resort to unity Playfab to store Player Data, and just authenticating with Steam and doing other things like getting friends. For now. I'd really have liked to be able to use the built in Steam features but it just doesn't seem secure enough right now. 

Thanks for the advice.

**Comment by DarklightStudios** [comment]
So I've come up with a slight workaround for Achievements... Since I'll be using PlayFab to store more secure things like stats, I'll also send an update to my Playfab player's internal data when an achievement is unlocked from the client. On start, I cycle through the list and check the steam achievements to make sure they match up. If they don't I'll reset all the player's Steam achievements either to match what's in the Playfab data OR to be petty, wipe them all.

I'm sure this could be done with stats and other things, but storing them on Steam would just be for cosmetic purposes and no game logic will actually use the Steam values for anything.

So now I'll look into the WebAPI and check if I can somehow trigger code automatically in Playfab when a steam achievement is unlocked on steam itself, automatically checking to see if it matches the internal data or not. So if anyone sets anything in Steam that didn't run through the client and pass the necessary checks required to also validate it into Playfab, it will automatically be undone OR they'll get wiped. I guess you could even go as far as to VAC ban them instantly, temp ban them through Playfab, give them a permanent "I tried to cheat Steam Achievements" achievement or what ever petty punishment you can think of...

**Comment by SupinePandora43** [comment]
> I guess you could even go as far as to VAC ban them instantly

VAC is only valve's thing.
i haven't seen it working (actually working) in other games than made by Valve.
and VAC doesn't ban for Achievements hacks.
maybe only when VAC secured game is running.

> give them a permanent "I tried to cheat Steam Achievements" achievement

you need to cheat Steam Achievements to get all Achievements lol

**Comment by DarklightStudios** [comment]
@SupinePandora43 Yeah maybe not VAC ban then, but you can ban players from Playfab at least.
And yeah giving an achievement for trying to cheat them might not be a good idea, can still give them a Cheater in-game title haha. That would be more for trying to cheat stats and other important things. :)
